Framework/Algorithms/src/LorentzCorrection.cpp

```diff
@@ -45,58 +68,99 @@ const std::string LorentzCorrection::name() const {
  */
 void LorentzCorrection::exec() {
-  MatrixWorkspace_sptr inWS = this->getProperty("InputWorkspace");
+  MatrixWorkspace_sptr inWS = this->getProperty(PropertyNames::INPUT_WKSP);
 
+  // clone the workspace - does nothing for inplace operation
   auto cloneAlg = this->createChildAlgorithm("CloneWorkspace", 0, 0.1);
   cloneAlg->initialize();
   cloneAlg->setProperty("InputWorkspace", inWS);
+  cloneAlg->setPropertyValue(
+      "OutputWorkspace", this->getPropertyValue(PropertyNames::OUTPUT_WKSP));
   cloneAlg->execute();
   Workspace_sptr temp = cloneAlg->getProperty("OutputWorkspace");
   MatrixWorkspace_sptr outWS =
       boost::dynamic_pointer_cast<MatrixWorkspace>(temp);
 
-  const auto numHistos = inWS->getNumberHistograms();
-  const auto &spectrumInfo = inWS->spectrumInfo();
-  Progress prog(this, 0.0, 1.0, numHistos);
+  Progress prog(this, 0.1, 1.0, inWS->getNumberHistograms());
+
+  const auto processingType = this->getPropertyValue(PropertyNames::TYPE);
+  if (processingType == TOF_SCD)
+    processTOF_SCD(outWS, prog);
+  else if (processingType == TOF_PD)
+    processTOF_PD(outWS, prog);
+  else {
+    std::stringstream msg;
+    msg << "Do not understand know how to process Type=\"" << processingType
+        << "\" - developer forgot to fill in if/else tree";
+    throw std::runtime_error(msg.str());
+  }
+
+  this->setProperty(PropertyNames::OUTPUT_WKSP, outWS);
+}
 
-  PARALLEL_FOR_IF(Kernel::threadSafe(*inWS))
-  for (int64_t i = 0; i < int64_t(numHistos); ++i) {
+void LorentzCorrection::processTOF_SCD(MatrixWorkspace_sptr &wksp,
+                                       Progress &prog) {
+  const int64_t numHistos = static_cast<int64_t>(wksp->getNumberHistograms());
+  const auto &spectrumInfo = wksp->spectrumInfo();
+
+  PARALLEL_FOR_IF(Kernel::threadSafe(*wksp))
+  for (int64_t i = 0; i < numHistos; ++i) {
     PARALLEL_START_INTERUPT_REGION
 
     if (!spectrumInfo.hasDetectors(i))
       continue;
 
-    const double twoTheta =
-        spectrumInfo.isMonitor(i) ? 0.0 : spectrumInfo.twoTheta(i);
-    const double sinTheta = std::sin(twoTheta / 2);
-    double sinThetaSq = sinTheta * sinTheta;
+    const double sinTheta_v = sinTheta(spectrumInfo, i);
+    const double sinThetaSq = sinTheta_v * sinTheta_v;
 
-    auto &inY = inWS->y(i);
-    auto &outY = outWS->mutableY(i);
-    auto &outE = outWS->mutableE(i);
-    const auto points = inWS->points(i);
+    auto &outY = wksp->mutableY(i);
+    auto &outE = wksp->mutableE(i);
+    const auto points = wksp->points(i);
+    const auto num_points = points.size();
     const auto pos = std::find(cbegin(points), cend(points), 0.0);
     if (pos != cend(points)) {
       std::stringstream buffer;
       buffer << "Cannot have zero values Wavelength. At workspace index: "
              << pos - cbegin(points);
       throw std::runtime_error(buffer.str());
     }
-    for (size_t j = 0; j < inY.size(); ++j) {
+    for (size_t j = 0; j < num_points; ++j) {
```

Comment:
Am I understanding the code correctly? Is this algorithm valid only for histogram data? Or did I miss an exec_event?